src/pudl/analysis/record_linkage/cross_year.py

```diff
@@ -0,0 +1,241 @@
+
+        arbitrary_types_allowed = True
+
+    def fit_predict(self, df: pd.DataFrame):
+        """Construct scikit-learn Pipeline and apply it to input data to assign IDs."""
+        if self.penalize_same_year:
+            distance_estimator = DistancePenalizeSameYear(
+                np.array(df.report_year), metric=self.distance_metric
+            )
+        else:
+            distance_estimator = PrecomputeDistance(metric=self.distance_metric)
+
+        for transform in self.column_transforms:
+            df = transform.clean_columns(df)
+
+        pipeline = Pipeline(
+            [
+                (
+                    "preprocess",
+                    ColumnTransformer(
+                        transformers=[
+                            transform.as_step() for transform in self.column_transforms
+                        ],
+                        transformer_weights={
+                            transform.step_name: transform.weight
+                            for transform in self.column_transforms
+                        },
+                    ),
+                ),
+                (
+                    "to_dense",
+                    DenseTransformer(),
+                ),
+                (
+                    "dim_reduction",
+                    PCA(copy=False, n_components=self.reduce_dims),
+                ),
+                ("precompute_dist", distance_estimator),
+                (
+                    "classifier",
+                    AgglomerativeClustering(
+                        n_clusters=self.n_clusters,
+                        metric="precomputed",
+                        linkage="average",
+                        distance_threshold=self.distance_threshold,
+                        compute_distances=True,
+                    ),
+                ),
+            ]
+        )
+
+        df[self.id_column] = pipeline.fit_predict(df)
+
+        return df
+
+
+class PrecomputeDistance(BaseEstimator, TransformerMixin):
+    """Precompute distances which are used during clustering and prediction."""
+
+    def __init__(self, metric="euclidean"):
+        """Initialize with distance metric."""
+        self.metric = metric
+
+    def fit(self, X, y=None, **fit_params):  # noqa: N803
+        """Required by scikit-learn, but does not modify anything."""
+        return self
+
+    def transform(self, X, y=None, **fit_params):  # noqa: N803
+        """Compute distance between records then add penalty to records from same year."""
+        return pairwise_distances(X, metric=self.metric)
+
+
+class DistancePenalizeSameYear(PrecomputeDistance):
+    """Custom estimator to compute distances used to identify clusters of plants."""
+
+    def __init__(self, report_years: np.array, metric="euclidean", penalty=100):
+        """Initialize estimator with configurable parameters.
+
+        Args:
+            report_years: reporty_year column used to penalize distance for records
```

Comment:
I feel confused about this `report_years` parameter, it's an array of all the records that share the same report year? I think some clarification in this docstring could be helpful

Reply:
I changed the comment a bit, let me know if it's any clearer. I made this just take the report year column instead of the whole dataframe, because I wasn't sure if the dataframe was going to be modified during the pipeline. We could try to change it back if you think that'd be clearer though